BlueprintUI/Sources/Element/ElementState.swift

```diff
@@ -55,39 +68,62 @@ final class ElementStateTree {
             return new
         }
 
-        if root == nil, let element = element {
-            /// Transition from no root element, to root element.
-            let new = makeRoot(with: element)
+        if let root = self.root {
+            if type(of: root.element.latest) == type(of: element) {
 
-            delegate.perform {
-                $0.tree(self, didSetupRootState: new)
-            }
-        } else if root != nil, element == nil {
-            /// Transition from a root element, to no root element.
-            let old = root
-            root = nil
+                root.prepareForLayout()
 
-            delegate.perform {
-                $0.tree(self, didTeardownRootState: old!)
-            }
-        } else if let root = self.root, let element = element {
-            if type(of: root.element.latest) == type(of: element) {
-                /// The type of the new element is the same, update inline.
                 root.update(with: element, in: environment, identifier: root.identifier)
 
-                delegate.perform {
+                let output = updates(root)
+
+                delegate.ifDebug {
                     $0.tree(self, didUpdateRootState: root)
                 }
+
+                root.finishedLayout()
```

Comment:
Related to my other comment, this recursive call gets called on all measurements. Can we optimize this to only run in certain conditions where we know we need to clean up? Looking at what it calls, we don't always skip when `wasVisited` is true, meaning it's recursive all the time.